Feature/contact shield support

minkiapps opened the pull request:
## Changes

Integrate HMS Contact Shield, introduce new dimension for HMS integration, introduce wrapper interface for GMS Exposure API and HMS Contact Shield API

Commit:
add modules 'gms' and 'hms'

Commit:
add nearby-api-aar file as aar-module (instead of direct aar dependency)

Commit:
add gms exposure client logic

Commit:
add flavor-dimension 'platform' for google (gms) and huawei (hms)

Commit:
add flavor-specifc dependencies for gms and hms

Commit:
remove obsolete nearby-api libs/aar-files (these are replaced by aar-module)

Commit:
add google-nearby-api module (instead of direct access of aar file via libs directory)

Commit:
add TODO for code improvements

Commit:
add play-services-nearby-exposurenotification-1.6.1-eap as AAR-module instead of directly accessing AAR-file via libs

Commit:
add Koin to hms module and provide koin-module that provides common-exposure-client (that is based on contactshield)

Commit:
add Koin to gms module and provide koin-module that provides common-exposure-client (that is based on google nearby)

Commit:
add flavor-specific koin-modules that either provides GMS or HMS classes based on hms/gms app-flavor

Commit:
downgrade from Koin 2.0.1 to 1.0.2 because app-layer still depends on 1.0.2 and updating koin would require further refactoring that is not related to ContactShield

Commit:
update koin imports

Commit:
extract nested god-class 'ExposureNotificationPhase' into it's own class file and update ExposureNotificationManager to use platform-specific exposure-client

Commit:
update platform-specific koin-module

Commit:
update ExposureNotificationRepository to use platform-specific exposure-client

Commit:
add agconnect-services.json to gitignore to prevent it from being added to version control

Commit:
apply plugin: 'com.huawei.agconnect'

Commit:
update DebugExposureNotificationsViewModel to use platform-specific exposure client

Commit:
add code that got lost during refactoring

Commit:
display huawei error cards on dashboard

Commit:
add further logs

Commit:
provide ContactShieldIntentService and pass its class to CommonExposureClient during dependency injection

Commit:
remove unwanted file from vcs

minkiapps committed:
update contact shield to 5.0.2, rework di for contact shield
fix bug where conversation of contact sketch to contact summary is wrong
use desugar core library to use Java 8 date API
don't show delete ID option in side menu on non GMS device

minkiapps committed:
update proguard file for hms dependencies

uliluckas commented:
First quick feedback:
https://github.com/austrianredcross/stopp-corona-android/pull/222/checks?check_run_id=1197641513

minkiapps committed:
adapt github action yml, build google and huawei variant

minkiapps committed:
hide system exposure setting button on non GMS phones

minkiapps committed:
map report type from hms model to gms

minkiapps committed:
don't expose TEK of today in hms implementation

minkiapps committed:
update contact shield to 5.0.3.300

minkiapps committed:
Update contact shield to 5.0.4 and remove "remove today TEK" logic

Latest update of contact shield doesn't expose today's TEK anymore
https://huawei-minki.slack.com/archives/D01BF0XBVMF/p1604651801000900

uliluckas commented on gms/src/main/java/at/roteskreuz/stopcorona/gms/GoogleExposureClient.kt:
```diff
@@ -0,0 +1,72 @@
+) : CommonExposureClient {
+
+    override suspend fun start() {
+        exposureNotificationClient.start().await()
```
Nice!

uliluckas commented on gms/src/main/java/at/roteskreuz/stopcorona/gms/GoogleExposureClient.kt:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    override suspend fun provideDiagnosisKeys(archives: List<File>, exposureConfiguration: ExposureConfiguration, token: String) {
+        exposureNotificationClient.provideDiagnosisKeys(archives, exposureConfiguration, token)
```
Please call `await()` to avoid returning early

minkiapps replied:
nice catch!

uliluckas commented on gms/src/main/java/at/roteskreuz/stopcorona/gms/GoogleExposureClient.kt:
```diff
@@ -0,0 +1,72 @@
+
+    override fun getServiceStatus(): ExposureServiceStatus {
+
+        val statusCode: Int = googleApiAvailability.isGooglePlayServicesAvailable(application)
```
No need for type declaration `: Int`

uliluckas commented on gms/src/main/java/at/roteskreuz/stopcorona/gms/GoogleExposureClient.kt:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    private companion object {
+        const val MIN_SUPPORTED_GOOGLE_PLAY_APK_VERSION = 201813000
```
~~Please clean up original, now unused constant.~~
Please remove `MIN_SUPPORTED_GOOGLE_PLAY_APK_VERSION` from gms module and see comment at use-site.
This constant is app specific. Not wrapper specific.

uliluckas commented on app/build.gradle:
```diff
@@ -3,6 +3,7 @@ apply plugin: 'kotlin-android'
 apply plugin: 'kotlin-android-extensions'
 apply plugin: 'kotlin-kapt'
 apply plugin: 'com.google.android.gms.oss-licenses-plugin'
+apply plugin: 'com.huawei.agconnect'
```
Please avoid loading the plugin for the google build. Might need a new module?

minkiapps replied:
apply plugin only in huawei build flavor

uliluckas commented on app/build.gradle:
```diff
@@ -135,6 +154,17 @@ dependencies {
+    // we keep the unreliable file dependency with manual play-services-base
+    implementation 'com.google.android.gms:play-services-base:17.3.0'
+    implementation project(":play-services-nearby-exposurenotification-1.6.1-eap")
+//    implementation project(":play-services-nearby-18.0.2-eap")
```
Please clean up todo and commented out code

uliluckas commented on app/src/main/AndroidManifest.xml:
```diff
@@ -83,5 +82,11 @@
+            android:name="at.roteskreuz.stopcorona.hms.ContactShieldIntentService"
+            android:enabled="true"
+            android:exported="true" />
+
```
Please move to HMS specific manifest and let gradle merge it

uliluckas commented on app/src/main/java/at/roteskreuz/stopcorona/model/managers/ExposureNotificationPhase.kt:
```diff
@@ -0,0 +1,624 @@
+ * needed for running.
+ */
+// TODO: Break-up god-class and move GMS- and HMS-only logic into corresponding platform-specific module
+sealed class ExposureNotificationPhase : ExposureServiceStatus {
```
Please do ;-)
At least the separation of GMS/HMS

minkiapps replied:
Well, this TODO from the original author is pretty ambitious. Before I can refactor that, we need to refactor a whole bunch of things before... I can only quote LOTR, where the [orc reported to Saruman](https://youtu.be/9qK7cweKy64?t=75): "the roots go deep".

minkiapps replied:
I think I could solve it with this [commit](https://github.com/austrianredcross/stopp-corona-android/pull/222/commits/9d45e2e43079b785d9427bd0806cc14bd3bec93f)

uliluckas commented on gms/src/main/java/at/roteskreuz/stopcorona/gms/GoogleExposureClient.kt:
```diff
@@ -0,0 +1,72 @@
+        val statusCode: Int = googleApiAvailability.isGooglePlayServicesAvailable(application)
+        val version = googleApiAvailability.getApkVersion(application)
+
+        if (version < MIN_SUPPORTED_GOOGLE_PLAY_APK_VERSION) {
```
Please inject this constant as it is app-specific, not wrapper specific.

uliluckas commented on app/src/main/java/at/roteskreuz/stopcorona/model/repositories/ExposureNotificationRepository.kt:
```diff
@@ -232,13 +247,13 @@ class ExposureNotificationRepositoryImpl(
     }
 
     override suspend fun refreshAndGetAppRegisteredForExposureNotificationsCurrentState(): Boolean {
-        val enabled = exposureNotificationClient.isEnabled.await()
```
Please remove dependency to `kotlinx-coroutines-play-services`

uliluckas commented on play-services-nearby-18.0.2-eap/build.gradle:
```diff
@@ -0,0 +1,2 @@
+configurations.maybeCreate("default")
+artifacts.add("default", file('play-services-nearby-18.0.2-eap.aar'))
```
Please remove the whole module and all commented out references

minkiapps committed:
Adapt debug menu to obtain GMS or HMS version name

uliluckas commented on common_exposure/build.gradle:
```diff
@@ -0,0 +1,44 @@
+    androidTestImplementation 'androidx.test.ext:junit:1.1.2'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.3.0'
+
+}
```
Please clean up depedencies. I immediately see `play-services-base` and `dir: 'libs'` but expect others to be unused as well.

minkiapps replied:
Done

uliluckas commented on common_exposure/src/androidTest/java/at/roteskreuz/stopcorona/commonexposure/ExampleInstrumentedTest.kt:
```diff
@@ -0,0 +1,24 @@
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+@RunWith(AndroidJUnit4::class)
+class ExampleInstrumentedTest {
```
Please remove examples.

uliluckas commented on hms/src/main/java/at/roteskreuz/stopcorona/hms/HuaweiExposureClient.kt:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    override suspend fun getTemporaryExposureKeys(): List<TemporaryExposureKey> {
+        val periodKeys = removeTEKsOfToday(contactShieldEngine.periodicKey.await())
```
Please adjust all properties of the TemporaryExposureKey to behave more like GMS. Especially `rollingPeriod` and `rollingStartIntervalNumber`

minkiapps replied:
done in this [commit](https://github.com/austrianredcross/stopp-corona-android/pull/222/commits/67b9960b1cdf460d92ae56335226b031bb841f4a)

uliluckas commented on build.gradle:
```diff
@@ -9,20 +9,28 @@ buildscript {
         classpath "com.github.triplet.gradle:play-publisher:2.1.0"
         classpath "com.github.marcpoppleton:git-changelog:0.1.3"
         classpath 'com.google.android.gms:oss-licenses-plugin:0.10.2'
+        classpath 'com.huawei.agconnect:agcp:1.3.2.301'
```
Please avoid applying the huawei plugin on the root level

minkiapps replied:
Unfortunately, I didn't find any proper solution to segregate build script class path, couldn't find any useful practise on internet. I think this should be fine, since Gradle will only download the agc plugin on a huawei build, since only huawei build flavor applies the agc plugin.

uliluckas commented on hms/build.gradle:
```diff
@@ -0,0 +1,64 @@
+    testImplementation 'junit:junit:4.13'
+    androidTestImplementation 'androidx.test.ext:junit:1.1.2'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.3.0'
+
```
Please clean up depedencies. I immediately see play-services-base and dir: 'libs' but expect others to be unused as well.

uliluckas commented on common_exposure/build.gradle:
```diff
@@ -0,0 +1,44 @@
+        release {
+            minifyEnabled false
+            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
+        }
```
minify is turned off by default. And I guess no proguard file is needed either. We never want to run proguard on a module. Only on the full app.

uliluckas commented on gms/build.gradle:
```diff
@@ -0,0 +1,50 @@
+        release {
+            minifyEnabled false
+            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
+        }
```
minify is turned off by default. And I guess no proguard file is needed either. We never want to run proguard on a module. Only on the full app.

uliluckas commented on hms/build.gradle:
```diff
@@ -0,0 +1,64 @@
+        release {
+            minifyEnabled false
+            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
+        }
```
minify is turned off by default. And I guess no proguard file is needed either. We never want to run proguard on a module. Only on the full app.

uliluckas commented on gms/build.gradle:
```diff
@@ -0,0 +1,50 @@
+        versionCode 1
+        versionName "1.0"
+
+        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
```
I am not sure if libraries can be instrumented?

uliluckas reviewed (changes requested):
First round of review.
Note to my self: Review refactoring of ExposureNotificationManager.kt

minkiapps committed:
Cleanup unused code and dependencies, move huawei specific stuff into hms module

minkiapps committed:
Inject min supported google play apk version from app module

minkiapps committed:
Correct parameters of period key to behave like Google

minkiapps committed:
Remove gms and hms module and move their code back to app module, segregate them through build variant folders

minkiapps committed:
Segregate hms specific error UI logic

minkiapps committed:
Add specific delay for contact shield autostart, so the permission activity is started not on top of RouterActivity

If exposure client is started on application start, it calls a permission activity, on the same time, RouterActivity starts
Dashboard Activity and overlaps the permission activity

minkiapps committed:
Fix bug where await() extension is not called in contactShieldEngine.putSharedKeyFiles

minkiapps committed:
Wrap debug output view in a scrollview

minkiapps committed:
Tweak CS processing debug notification, ignore gms specific business logic for delayed processing

minkiapps committed:
Fix KeyboardHelper callback logic

minkiapps committed:
Add debug preference to track start time of keymatch processing

Commit:
Prepare for release 2.0.6

minkiapps committed:
Add signing config for hms release variant

minkiapps committed:
Fix logic where delayed key processing should be only applied to GMS

Plant Timber debug tree if app is debuggable

minkiapps committed:
Add exception description for debug build in error dialog

minkiapps committed:
Merge branch 'task/prepare_for_release_2.0.6' into feature/contact_shield_support

minkiapps committed:
Clamp CS exposure values to GMS allowed value range

minkiapps committed:
Change faq link

minkiapps committed:
Let bluetooth manager emit initial bluetooth enabled state when start listening for changes

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support

minkiapps committed:
Add hms crashlytics

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support

minkiapps committed:
Add Huawei AppGallery link in share text

minkiapps committed:
Add Huawei UI text in change log

minkiapps committed:
Update privacy and terms of use

minkiapps committed:
Fix change log text

minkiapps committed:
Update German privacy text

minkiapps committed:
Add error handling if user does not allow permanent location access

minkiapps committed:
Add small hack to trigger an update window if hms core has no nearby kit installed and return error code 1212

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support

# Conflicts:
#	app/src/main/java/at/roteskreuz/stopcorona/screens/savedIDs/InfoDeleteExposureKeysFragment.kt
#	app/src/main/java/at/roteskreuz/stopcorona/utils/KeyboardHelper.kt

minkiapps committed:
Update privacy english version

minkiapps committed:
Add intent to hms system contact shield page

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support

minkiapps committed:
Update to latest contact shield version

minkiapps committed:
Handle framework specific error on upload exposure keys

fraiss-devteam committed:
Fix update screen visibility

alexritz commented on app/src/google/java/at/roteskreuz/stopcorona/di/ExposureModule.kt:
```diff
@@ -0,0 +1,18 @@
+    single { Nearby.getExposureNotificationClient(androidApplication()) }
+
+    single<CommonExposureClient> {
+        GoogleExposureClient(androidApplication(), get(), get())
```
Please provide naming for get() calls

alexritz commented on app/src/huawei/java/at/roteskreuz/stopcorona/di/ExposureModule.kt:
```diff
@@ -0,0 +1,22 @@
+    single { ContactShield.getContactShieldEngine(androidApplication()) }
+
+    single<CommonExposureClient> {
+        HuaweiExposureClient(androidApplication(), get(), get(), androidApplication().getSharedPreferences(PREFS_HMS_DEBUG, Context.MODE_PRIVATE))
```
Please provide naming for get() calls

alexritz commented on app/src/google/java/at/roteskreuz/stopcorona/screens/dashboard/epoxy/DashBoardController.kt:
```diff
@@ -0,0 +1,112 @@
+package at.roteskreuz.stopcorona.screens.dashboard.epoxy
```
Please fix typo in the filename: DashBoardController ->DashboardController

alexritz commented on app/src/main/AndroidManifest.xml:
```diff
@@ -1,6 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
-<manifest
-    xmlns:android="http://schemas.android.com/apk/res/android"
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
```
Please remove code reformatting commit

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support

# Conflicts:
#	app/src/main/assets/privacy-and-terms-of-use_de.html
#	app/src/main/assets/privacy-and-terms-of-use_en.html
#	app/src/main/assets/privacy_de.html
#	app/src/main/assets/privacy_en.html

minkiapps committed:
Handle force update and redirect to platform specific app store

minkiapps committed:
Merge branch 'feature/contact_shield_support' into hot_fix_release

# Conflicts:
#	app/src/main/java/at/roteskreuz/stopcorona/di/ViewModelModule.kt
#	app/src/main/java/at/roteskreuz/stopcorona/screens/dashboard/DashboardFragment.kt
#	app/src/main/java/at/roteskreuz/stopcorona/screens/dashboard/DashboardViewModel.kt
#	app/src/main/java/at/roteskreuz/stopcorona/screens/mandatory_update/MandatoryUpdateFragment.kt

minkiapps committed:
Merge pull request #1 from minkiapps/hot_fix_release

Hot fix release

minkiapps committed:
Restore ignore battery optimization logic

minkiapps committed:
Incorporate PR review comments

minkiapps committed:
Update contact shield, adapt update kit handling

minkiapps committed:
Tweak namings

minkiapps committed:
Refresh state after core update was successful

minkiapps committed:
Update contact shield version

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support

minkiapps committed:
Update agcp classpath

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support

# Conflicts:
#	app/src/main/java/at/roteskreuz/stopcorona/screens/debug/diagnosis_keys/DebugDiagnosisKeysViewModel.kt

minkiapps committed:
Update agcp to support multiple flavour in one agconnect json file

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support

minkiapps committed:
Add dark mode support for webview on HMS phones

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support

# Conflicts:
#	app/src/main/java/at/roteskreuz/stopcorona/screens/webView/WebViewWithAssetsResourcesFragment.kt

minkiapps committed:
Merge branch 'develop' into feature/contact_shield_support, update dependencies and pugins

# Conflicts:
#	app/build.gradle
#	app/src/main/java/at/roteskreuz/stopcorona/screens/dashboard/DashboardFragment.kt

minkiapps committed:
Resolve cyclic dependencies which causes a stackoverflow error in gradle task app:getDependencies

Ein-Tim commented:
I doubt that this PR will be merged, but if you want to you could provide a PR for the German Corona-Warn-App: https://github.com/corona-warn-app
But before doing this, it should certainly be discussed whether the RKI is open for publishing the app on the Huawei App Gallery.
